File: javascript/triangle/triangle.js
const typesList = [
  {
    type: 'illegal',
    check: (a, b, c) => (a <= 0 || a + b < c),
  },
  {
    type: 'equilateral',
    check: (a, b, c) => (a === b && b === c),
  },
  {
    type: 'isosceles',
    check: (a, b, c) => (a === b || b === c),
  },
  {
    type: 'scalene',
    check: (a, b, c) => (a !== b && b !== c),
  },
];

const checker = (sides) => {
  const [a, b, c] = sides.sort((x, y) => x - y);
  const { type } = typesList.find(el => el.check(a, b, c));

  if (type === 'illegal') {
    throw new Error();
  }

  return type;
};

class Triangle {
  constructor(a, b, c) {
    this.sides = [a, b, c];
  }

  kind() {
    return checker(this.sides);
  }
}

export default Triangle;

File: erlang/triangle/src/triangle.erl
-module(triangle).

-export([kind/3]).


kind(0, 0, 0) -> {error, "all side lengths must be positive"};

kind(A, B, C) when A > B + C; B > A + C; C > A + B ->
  {error, "side lengths violate triangle inequality"};

kind(A, A, A) -> equilateral;
kind(A, A, _) -> isosceles;
kind(_, A, A) -> isosceles;
kind(A, _, A) -> isosceles;
kind(_, _, _) -> scalene.
